File: app/tests/test_mobile_playback_video_cards.py
"""Mobile Playback/Events video-first redesign (2026-09-04): source-
structure tests for renderMobileRecentEvents()'s new one-large-card-
per-row markup, and the root-cause fix it ships alongside (an event's
own event.thumbnail is now used directly instead of findClipNear()
guessing a nearby recording's thumbnail -- see that function's own
comment in main.py for the full root-cause trace).

The companion .mjs file (test_mobile_playback_video_cards.mjs,
alongside this one) executes the ACTUAL extracted function via Node,
with a stubbed findClipNear() that throws if ever called -- that is
the real behavioral proof this suite's own established idiom (see
test_playback_analytics_lanes.py's own docstring) calls for. This file
proves everything a plain render + string assertions on the real,
unmodified _render_customer_playback() output can prove without
executing JS: the new CSS classes exist, the old plain-text list rows
are gone, findClipNear() is genuinely absent from this one function's
own source (not just untriggered in a particular test scenario), the
existing playClip()/event-media-url()/revealClipPanel() wiring is
byte-for-byte unchanged, and the desktop timeline/lane code this
change must never touch is still there.

Same import-inside-container constraint and _fake_request()/
_render_customer_playback() calling convention as this suite's other
Playback tests.
"""

from types import SimpleNamespace

import main


def _fake_request(t=None, camera=None):
    return SimpleNamespace(query_params=SimpleNamespace(get=lambda key, default=None: {"t": t, "camera": camera}.get(key, default)))


def _render(monkeypatch, recordings=None, events=None):
    monkeypatch.setattr(main, "_customer_recording_rows", lambda camera_id, **kwargs: recordings or [])
    monkeypatch.setattr(main, "_customer_detection_events", lambda request: events or [])
    monkeypatch.setattr(main, "_customer_camera_events", lambda camera_id, date: events or [])
    return main._render_customer_playback([{"id": "cam-1", "name": "Front Door", "camera_number": 1}], _fake_request())


def _mobile_function_body(html):
    start = html.index("function renderMobileRecentEvents(cameraId,clips,events)")
    end = html.index("// === LANE_CORE_START ===", start)
    assert end > start
    return html[start:end]


def _active_code_only(body):
    """Strips '//...' line comments -- for assertions that must hold
    against the real, executable code, not against this function's own
    explanatory comments (which legitimately quote the old buggy
    findClipNear()/"No preview" behavior by name, for documentation,
    exactly like the rest of this codebase's own commenting
    convention). The real behavioral proof that findClipNear() is never
    actually invoked is test_mobile_playback_video_cards.mjs, which
    executes this exact source for real; this is a lighter-weight,
    still-real structural cross-check on top of that, not a
    replacement for it."""
    return "\n".join(line for line in body.split("\n") if "//" not in line.strip()[:2])


# ---------------------------------------------------------------------------
# 1. New card markup/CSS exists.
# ---------------------------------------------------------------------------

def test_mobile_media_card_css_classes_are_defined(monkeypatch):
    html = _render(monkeypatch)
    for selector in (
        ".mobile-media-card{", ".mobile-media-thumb{", ".mobile-media-time{",
        ".mobile-media-badge{", ".mobile-media-menu{", ".mobile-media-fallback{",
    ):
        assert selector in html, f"missing CSS rule for {selector}"


def test_mobile_media_card_css_scoped_to_the_existing_mobile_only_toggle(monkeypatch):
    # .mobile-recent-events (the section these cards live in) is already
    # hidden above 900px by the pre-existing media query -- the new card
    # rules deliberately don't need (and don't have) a second breakpoint
    # of their own, since their whole container is already conditional.
    html = _render(monkeypatch)
    toggle_idx = html.index("@media (max-width:900px){.monitor-timeline{display:none!important}.mobile-recent-events{display:block!important}}")
    cards_idx = html.index(".mobile-media-card{")
    assert toggle_idx < cards_idx


# ---------------------------------------------------------------------------
# 2. The old plain-text list rows are gone.
# ---------------------------------------------------------------------------

def test_old_health_row_mobile_card_markup_is_removed(monkeypatch):
    html = _render(monkeypatch)
    body = _mobile_function_body(html)
    assert 'class="health-row" data-mobile-clip' not in body
    assert 'class="health-row" data-mobile-event' not in body
    assert "<span class=\"health-name\">" not in body
    assert ">Clip</span>" not in body


def test_no_preview_and_analytics_only_text_are_gone_from_this_function(monkeypatch):
    html = _render(monkeypatch)
    code = _active_code_only(_mobile_function_body(html))
    assert "No preview" not in code
    assert "Analytics only" not in code
    assert "No clip available" in code  # the new compact fallback, still present as a template string


# ---------------------------------------------------------------------------
# 3. Root-cause fix: event.thumbnail is used, findClipNear() is not.
# ---------------------------------------------------------------------------

def test_event_thumbnail_field_is_used_directly(monkeypatch):
    html = _render(monkeypatch)
    body = _mobile_function_body(html)
    assert "event.thumbnail" in body


def test_findclipnear_is_completely_absent_from_this_functions_actual_code(monkeypatch):
    # Not just untriggered in one test scenario -- see this same claim
    # proven behaviorally (a throwing stub, never invoked across 5
    # scenarios) in test_mobile_playback_video_cards.mjs. This is a
    # second, independent structural check on top of that: the literal
    # call is gone from the real code, not merely dead-code-eliminated
    # at runtime. Comments are excluded (see _active_code_only()) since
    # this function's own explanatory comments legitimately name
    # findClipNear() when documenting the bug they fixed.
    html = _render(monkeypatch)
    code = _active_code_only(_mobile_function_body(html))
    assert "findClipNear" not in code


def test_hydratepreview_lazy_lookup_is_removed(monkeypatch):
    # The old lazy, findClipNear()-based thumbnail hydration is no
    # longer needed at all now that event.thumbnail is already correct
    # and present at initial render time.
    html = _render(monkeypatch)
    body = _mobile_function_body(html)
    assert "hydratePreview" not in body


# ---------------------------------------------------------------------------
# 4. Existing playback wiring is byte-for-byte unchanged.
# ---------------------------------------------------------------------------

def test_recording_card_still_calls_the_existing_unmodified_playclip(monkeypatch):
    html = _render(monkeypatch)
    body = _mobile_function_body(html)
    assert "row.addEventListener('click',()=>playClip(cameraId,clip));" in body


def test_event_card_still_uses_the_existing_unmodified_event_media_url_and_reveal_panel(monkeypatch):
    html = _render(monkeypatch)
    body = _mobile_function_body(html)
    assert "/api/customer/events/${cameraId}/${eventId}/media/url" in body  # rendered output, single braces (f-string already evaluated)
    assert "revealClipPanel();" in body
    assert "video.src=payload.url;" in body


def test_analytics_only_events_are_still_not_wired_as_playable_controls(monkeypatch):
    html = _render(monkeypatch)
    body = _mobile_function_body(html)
    assert "if(row.dataset.mobileEventClip!=='1'||!row.dataset.mobileEventId)return;" in body


# ---------------------------------------------------------------------------
# 5. Desktop timeline/lane code is completely untouched.
# ---------------------------------------------------------------------------

def test_desktop_lane_and_timeline_code_still_present_unchanged(monkeypatch):
    html = _render(monkeypatch)
    assert "const EVENT_LANE_ORDER=['motion','person','vehicle','lpr','people_counting','intrusion'];" in html
    assert '<div class="timeline-lane" id="playback-timeline-lane">' in html
    assert 'class="monitor-timeline"' in html


def test_desktop_vs_mobile_toggle_css_itself_is_unchanged(monkeypatch):
    html = _render(monkeypatch)
    assert (
        "<style>@media (max-width:900px){.monitor-timeline{display:none!important}"
        ".mobile-recent-events{display:block!important}}@media (min-width:901px)"
        "{.mobile-recent-events{display:none!important}.monitor-timeline{display:block}}</style>"
    ) in html

File: app/tests/test_mobile_playback_video_cards.mjs
#!/usr/bin/env node
// Real execution (not string assertions) of the ACTUAL shipped
// renderMobileRecentEvents() card-building logic, extracted directly
// from main.py between its own function boundary and the
// mobileList.innerHTML=... assignment (see the extraction in this
// suite's own runner script) -- same idiom this project already
// established in test_playback_analytics_lanes.mjs/test_playback_
// segment_chaining_core.mjs for pure/DOM-light front-end logic.
//
// Proves, against the real code, not a paraphrase of it:
//   - an event with a real event.thumbnail renders that exact URL
//   - findClipNear() is never called anywhere in this function (the
//     stub below throws if it ever is -- this is the actual regression
//     test for the root-cause bug: the old code called findClipNear()
//     to guess an event's preview from a nearby recording instead of
//     using the event's own already-correct thumbnail field)
//   - an event with no thumbnail gets the compact "<Type> · No clip
//     available" fallback, never a large empty box
//   - a recording (clip) row still renders from clip.id, independent
//     of any of the above
//
// Usage: node test_mobile_playback_video_cards.mjs <path-to-main.py>

import assert from "node:assert/strict";
import fs from "node:fs";

const mainPyPath = process.argv[2];
if (!mainPyPath) {
  console.error("usage: node test_mobile_playback_video_cards.mjs <path-to-main.py>");
  process.exit(2);
}

const src = fs.readFileSync(mainPyPath, "utf8");
const startMarker = "  function renderMobileRecentEvents(cameraId,clips,events){{";
// Two separate single-line needles (not one multi-line string) so this
// survives either LF or CRLF line endings in main.py.
const endNeedleLine1 = "mobileList.innerHTML=(clipRows+eventRows)||";
const endNeedleLine2 = "No recent activity for this camera.</div>';";

const start = src.indexOf(startMarker);
if (start === -1) throw new Error("renderMobileRecentEvents() not found in main.py");
const line1Idx = src.indexOf(endNeedleLine1, start);
if (line1Idx === -1) throw new Error("mobileList.innerHTML assignment not found after renderMobileRecentEvents()");
const line2Idx = src.indexOf(endNeedleLine2, line1Idx);
if (line2Idx === -1) throw new Error("end-of-function fallback text not found after the innerHTML assignment");
const end = line2Idx + endNeedleLine2.length;

const jsSource =
  src
    .slice(start, end)
    .replaceAll("{{", "{")
    .replaceAll("}}", "}") +
  "\n  }"; // closes the function body -- extraction deliberately stops
           // mid-body, right after the innerHTML assignment, before the
           // click-wiring code that needs a real DOM's querySelectorAll.

let passed = 0;
function ok(label) {
  passed++;
  console.log(`ok - ${label}`);
}

function run(scenarioName, { clips, events, filterCategoryImpl }) {
  const mobileListEl = { innerHTML: null };
  const documentStub = {
    getElementById(id) {
      if (id === "mobile-recent-events-list") return mobileListEl;
      return null;
    },
  };
  const EVENT_COLORS = {
    motion: "#f0b94d", person: "#4d9ef0", vehicle: "#a06df0",
    lpr: "#3dbfae", people_counting: "#4dcf7a", intrusion: "#f0554d",
  };
  const activeFilters = new Set(["motion", "person", "vehicle", "lpr", "people_counting", "intrusion"]);
  function filterCategory(eventType) {
    return filterCategoryImpl ? filterCategoryImpl(eventType) : (EVENT_COLORS[eventType] ? eventType : eventType);
  }
  function playbackDate(value) {
    return new Date(typeof value === "number" ? value : String(value) + (String(value).endsWith("Z") ? "" : "Z"));
  }
  function findClipNear() {
    throw new Error(`findClipNear() must never be called by renderMobileRecentEvents() (scenario: ${scenarioName})`);
  }

  const fn = new Function(
    "document", "EVENT_COLORS", "activeFilters", "filterCategory", "playbackDate", "findClipNear",
    jsSource + "\nreturn renderMobileRecentEvents;"
  )(documentStub, EVENT_COLORS, activeFilters, filterCategory, playbackDate, findClipNear);

  fn("cam-1", clips, events);
  return mobileListEl.innerHTML;
}

// --------------------------------------------------------------- an event with a real thumbnail

{
  const html = run("thumbnail-present", {
    clips: [],
    events: [{ id: "evt-1", event_type: "person", timestamp: "2026-09-04T18:00:00Z", has_event_clip: true, thumbnail: "/api/customer/events/cam-1/evt-1/thumbnail" }],
  });
  assert.match(html, /<img class="mobile-media-thumb" src="\/api\/customer\/events\/cam-1\/evt-1\/thumbnail"/);
  assert.doesNotMatch(html, /No clip available/);
  assert.match(html, /mobile-media-badge[^>]*>Person</);
  ok("an event with event.thumbnail renders that exact thumbnail URL");
}

// --------------------------------------------------------------- findClipNear() never called

{
  // The stubbed findClipNear() above throws immediately if invoked --
  // reaching this line at all (for a scenario that, under the old
  // buggy code, WOULD have called it for every single event row) is
  // itself the proof.
  run("findClipNear-guard", {
    clips: [{ id: "rec-1", start: "2026-09-04T17:55:00Z", end: "2026-09-04T18:00:00Z" }],
    events: [
      { id: "evt-1", event_type: "person", timestamp: "2026-09-04T18:00:00Z", has_event_clip: true, thumbnail: "/api/customer/events/cam-1/evt-1/thumbnail" },
      { id: "evt-2", event_type: "vehicle", timestamp: "2026-09-04T18:05:00Z", has_event_clip: false, thumbnail: null },
    ],
  });
  ok("findClipNear() is never called as the event-thumbnail source, with or without a nearby recording present");
}

// --------------------------------------------------------------- no media -> compact fallback, not a big empty box

{
  const html = run("no-thumbnail", {
    clips: [],
    events: [{ id: "evt-2", event_type: "vehicle", timestamp: "2026-09-04T18:05:00Z", has_event_clip: false, thumbnail: null }],
  });
  assert.match(html, /mobile-media-fallback">Vehicle · No clip available</);
  assert.doesNotMatch(html, /<img/);
  assert.doesNotMatch(html, /No preview/);
  assert.doesNotMatch(html, /Analytics only/);
  ok("an event genuinely without media gets the compact fallback, not a large empty preview box");
}

// --------------------------------------------------------------- recording (clip) row still renders

{
  const html = run("clip-row", {
    clips: [{ id: "rec-42", start: "2026-09-04T18:00:00Z", end: "2026-09-04T18:05:00Z" }],
    events: [],
  });
  assert.match(html, /data-mobile-clip role="button" tabindex="0"/);
  assert.match(html, /src="\/api\/customer\/recordings\/cam-1\/rec-42\/thumbnail"/);
  assert.match(html, /5 min/);
  ok("a recording still renders its own thumbnail card with a real duration");
}

// --------------------------------------------------------------- unfiltered/unsupported categories don't crash and don't call findClipNear either

{
  run("unsupported-category", {
    clips: [],
    events: [{ id: "evt-3", event_type: "smart_motion", timestamp: "2026-09-04T18:10:00Z", has_event_clip: false, thumbnail: null }],
  });
  ok("an unsupported/filtered-out category is handled without ever reaching findClipNear() either");
}

console.log(`\n${passed}/${passed} passed`);
